### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by jipspel**
- Improve implementation of fixed choices for MDPs
- fix several warnings

### CODE REVIEW

File: src/storm/solver/helper/OptimisticValueIterationHelper.cpp [hunk 1]
```diff
@@ -52,29 +52,34 @@ namespace storm {
...
                     STORM_LOG_ASSERT(x.size() > 0, "Empty equation system not expected.");
                     ValueType diff = storm::utility::zero<ValueType>();
                     
                     IndexType i = x.size();
                     while (i > 0) {
                         --i;
-                        ValueType newXi = HasRowGroups ? multiplyRowGroup<Dir>(i, b, x) : multiplyRow(i, b[i], x);
+                        ValueType newXi;
+                        if (schedulerFixedForRowgroup && schedulerFixedForRowgroup.get()[i]) {
+                            newXi = multiplyRow(i, b[i + scheduler.get().at(i)], x);
```
Comment (tquatmann) [inline src/storm/solver/helper/OptimisticValueIterationHelper.cpp:79]: This currently multiplies the `i`th row, right? shouldn't it be something like
```suggestion
                            auto row = i + scheduler.get().at(i);
                            newXi = multiplyRow(row, b[row], x);
```
same applies to similar occurrences below.

File: src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp [hunk 2]
```diff
@@ -45,6 +45,7 @@ namespace storm {
 
         template<typename ValueType>
         bool TopologicalMinMaxLinearEquationSolver<ValueType>::internalSolveEquations(Environment const& env, OptimizationDirection dir, std::vector<ValueType>& x, std::vector<ValueType> const& b) const {
+            // Only for this one we do something with the fixed choices
```
Comment (tquatmann) [inline src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp:48]: This comment seems to be a bit out of context

File: src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp [hunk 3]
```diff
@@ -240,36 +271,53 @@ namespace storm {
...
-            if (this->choiceFixedForState) {
-                // convert fixed states to only fixed states of sccs
+            this->sccSolver->setHasUniqueSolution(this->hasUniqueSolution());
+            this->sccSolver->setHasNoEndComponents(this->hasNoEndComponents());
+            this->sccSolver->setTrackScheduler(this->isTrackSchedulerSet());
+
+            storm::storage::SparseMatrix<ValueType> sccA;
+            if (this->choiceFixedForRowGroup) {
+                // Obtain choiceFixedForState bitvector containing only the states of the scc.
                 storm::storage::BitVector choiceFixedForStateSCC(sccRowGroups.getNumberOfSetBits());
```
Comment (tquatmann) [inline src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp:286]: This is shorter compared to the computation below:
```suggestion
                storm::storage::BitVector choiceFixedForStateSCC = this->choicesFixedForRowGroup % sccRowGroups;
```
Although I think `choiceFixedForStateSCC` is not used? Maybe there is a typo below.

File: src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp [hunk 4]
```diff
@@ -240,36 +271,53 @@ namespace storm {
...
+                // initial scheduler
+                if (this->hasInitialScheduler()) {
+                    auto sccInitChoices = storm::utility::vector::filterVector(this->getInitialScheduler(), sccRowGroups);
+                    // As we removed the entries where the choice was fixed, we need to change the scheduler.
+                    // We set the scheduler to 0 for those states.
+                    if (this->choiceFixedForRowGroup) {
+                        for (auto const choice : this->choiceFixedForRowGroup.get()) {
+                            sccInitChoices[choice] = 0;
+                        }
+                    }
```
Comment (tquatmann) [inline src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp:304]: Use `storm::utility::vector::setVectorValues` ?

Comment (tquatmann) [reply src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp:304]: And I think here is the typo I mentioned before. It should refer to `choiceFixedForStateSCC`, right?

**Review by tquatmann — CHANGES_REQUESTED** [review]
I haven't looked through the monotonicity details with great care, but the fixes there seem valid.

I'm going to run some  benchmarks to see if the changes in svi and ovi affect performance.

File: src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp [hunk 5]
```diff
@@ -64,6 +64,7 @@ namespace storm {
             bool result = false;
             switch (getMethod(env, storm::NumberTraits<ValueType>::IsExact || env.solver().isForceExact())) {
                 case MinMaxMethod::ValueIteration:
+                    // Throws error when choices for scheduler are fixed
```
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp:67]: I'll remove this comment as it is incorrect.

File: src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp [hunk 6]
```diff
@@ -76,6 +77,7 @@ namespace storm {
                     result = solveEquationsRationalSearch(env, dir, x, b);
                     break;
                 case MinMaxMethod::IntervalIteration:
+                    // Throws error when choices for scheduler are fixed
```
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp:80]: I'll remove the comment, however solveEquationsIntervalIteration does throw a NotImplementedException.

File: src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp [hunk 7]
```diff
@@ -318,23 +313,35 @@ namespace storm {
...
+
             SolverStatus status = SolverStatus::InProgress;
             while (status == SolverStatus::InProgress) {
-                // Compute x' = min/max(A*x + b).
-                if (useGaussSeidelMultiplication) {
-                    // Copy over the current vector so we can modify it in-place.
-                    *newX = *currentX;
-                    multiplier.multiplyAndReduceGaussSeidel(env, dir, *newX, &b);
+
+                // TODO: integrate fixed choices for the scheduler for a rowgroup in multiplier
```
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/solver/IterativeMinMaxLinearEquationSolver.cpp:320]: added a storm_log_throw

File: src/storm/storage/jani/eliminator/FunctionEliminator.cpp [hunk 8]
```diff
@@ -21,393 +21,391 @@ namespace storm {
```
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/storage/jani/eliminator/FunctionEliminator.cpp:21]: uhhhh no idea I'll look into it

File: src/storm/solver/helper/SoundValueIterationHelper.cpp [hunk 9]
```diff
@@ -111,86 +113,130 @@ namespace storm {
...
-                                // Check if the current row is better then the previously found one
-                                if (better<dir>(currentValue, bestValue)) {
-                                    xBest = std::move(xi);
-                                    yBest = std::move(yi);
-                                    bestValue = std::move(currentValue);
-                                } else if (currentValue == bestValue && yBest > yi) {
-                                    // If the value for this row is not strictly better, it might still be equal and have a better y value
-                                    xBest = std::move(xi);
-                                    yBest = std::move(yi);
+                        auto rowGroupIndex = groupStartIt - rowGroupIndices->rbegin() - 1;
```
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/solver/helper/SoundValueIterationHelper.cpp:130]: I don't understand the problem. We calculate the rowGroupIndex (which is the state number?) and then pick the correct row based on the current groupStart and the choice for the statenumber.

[the remark this answers is not in the document]
Comment (tquatmann) [reply src/storm/solver/helper/SoundValueIterationHelper.cpp:130]: When there are `n` states with indices `0, 1, ..., n-1`, the iteration (i.e. `groupStartIt`) refers to state `n-1` first, then `n-2`, ... all the way down to state 0 in the last iteration.  Notice that we have the reverse iterators via `rbegin()` (and not `begin()` here.

However, if I see this correctly, the `rowGroupIndex` starts with `0` in the first iteration and then counts upwards to `n-1` in the last iteration.

Is it clearer now? It's probably better to just initialize rowGroupIndex with `n` and do `--rowGroupIndex;` at the beginning of each iteration. Or I am completely wrong here 😅

File: src/storm/solver/LpMinMaxLinearEquationSolver.cpp [hunk 10]
```diff
@@ -65,9 +65,11 @@ namespace storm {
...
             for (uint64_t rowGroup = 0; rowGroup < this->A->getRowGroupCount(); ++rowGroup) {
                 // The rowgroup refers to the state number
-                if (this->choiceFixedForRowGroup && this->choiceFixedForRowGroup.get()[rowGroup]) {
-                    // The choice is fixed
-                    auto rowIndex = this->A->getRowGroupIndices()[rowGroup] + this->getInitialScheduler()[rowGroup];
+                for (uint64_t rowIndex = this->A->getRowGroupIndices()[rowGroup]; rowIndex < this->A->getRowGroupIndices()[rowGroup + 1]; ++rowIndex) {
+                    if (this->choiceFixedForRowGroup && this->choiceFixedForRowGroup.get()[rowGroup]) {
+                        // The choice is fixed so we set the rowIndex to the fixed one
+                        auto rowIndex = this->A->getRowGroupIndices()[rowGroup] + this->getInitialScheduler()[rowGroup];
+                    }
```
Comment (tquatmann) [inline src/storm/solver/LpMinMaxLinearEquationSolver.cpp]: The `auto` in line 71 creates a separate variable `rowIndex` that runs out of scope in line 72. I guess you wanted to set the `rowIndex` variable from line 68 instead? Anyway, I'd suggest to do it like this:
```suggestion
                uint64_t rowIndex, rowGroupEnd;
                if (this->choiceFixedForRowGroup && this->choiceFixedForRowGroup.get()[rowGroup]) {
                    rowIndex = this->A->getRowGroupIndices()[rowGroup] + this->getInitialScheduler()[rowGroup];
                    rowGroupEnd = rowIndex + 1;
                } else {
                    rowIndex = this->A->getRowGroupIndices()[rowGroup];
                    rowGroupEnd = this->A->getRowGroupIndices()[rowGroup + 1];
                }
                for (; rowIndex < rowGroupEnd; ++rowIndex) {
```
Then, the `if` block in line 87 below can be removed and we check for fixed choices less often.

File: src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp [hunk 11]
```diff
@@ -283,27 +282,16 @@ namespace storm {
...
                 if (this->hasInitialScheduler()) {
                     auto sccInitChoices = storm::utility::vector::filterVector(this->getInitialScheduler(), sccRowGroups);
                     // As we removed the entries where the choice was fixed, we need to change the scheduler.
                     // We set the scheduler to 0 for those states.
-                    if (this->choiceFixedForRowGroup) {
-                        for (auto const choice : this->choiceFixedForRowGroup.get()) {
-                            sccInitChoices[choice] = 0;
-                        }
-                    }
+                    storm::utility::vector::setVectorValues(sccInitChoices, this->choiceFixedForStateScc.get(), 0);
```
Comment (tquatmann) [inline src/storm/solver/TopologicalMinMaxLinearEquationSolver.cpp:293]:
```suggestion
                    storm::utility::vector::setVectorValues(sccInitChoices, choiceFixedForStateScc, 0);
```

File: src/storm/solver/helper/OptimisticValueIterationHelper.cpp [hunk 12]
```diff
@@ -141,14 +154,19 @@ namespace storm {
                         // we now know that we did not converge. We still need to set the remaining values
                         while (i > 0) {
                             --i;
-                            x[i] = HasRowGroups ? multiplyRowGroup<Dir>(i, b, x) : multiplyRow(i, b[i], x);
+                            if (schedulerFixedForRowgroup && schedulerFixedForRowgroup.get()[i]) {
+                                x[i] = multiplyRow(i, b[i + scheduler.get().at(i)], x);
```
Comment (tquatmann) [inline src/storm/solver/helper/OptimisticValueIterationHelper.cpp:158]:
```suggestion
                                auto row = i + scheduler.get().at(i);
                                newXi = multiplyRow(row, b[row], x);
```

**Review by tquatmann — CHANGES_REQUESTED** [review]

File: src/storm/solver/helper/SoundValueIterationHelper.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jipspel) [reply src/storm/solver/helper/SoundValueIterationHelper.cpp:130]: ahh yeah now I get it :)

**Review by tquatmann — APPROVED** [review]

**Pull request merged**